In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/cars-formated/cars_formated.csv


In [3]:
# uploading dataset
db_source = '/kaggle/input/cars-formated/cars_formated.csv'
# puting data base to the 
cars_formated = pd.read_csv(db_source, index_col = 'Unnamed: 0')

In [4]:
cars_formated

year    make                model       body transmission  \
0       2015     kia              sorento        suv    automatic   
1       2015     kia              sorento        suv    automatic   
2       2014     bmw             3 series      sedan    automatic   
3       2015   volvo                  s60      sedan    automatic   
4       2014     bmw  6 series gran coupe      sedan    automatic   
...      ...     ...                  ...        ...          ...   
558832  2015     kia                 k900      sedan    automatic   
558833  2012     ram                 2500   crew cab    automatic   
558834  2012     bmw                   x5        suv    automatic   
558835  2015  nissan               altima      sedan    automatic   
558836  2014    ford                f-150  supercrew    automatic   

                      vin  condition  odometer   color  sellingprice  
0       5xyktca69fg566472        5.0   16639.0   white       21500.0  
1       5xyktca69fg561319        5.0    9393.0   white       21500.0  
2       wba3c1c51ek116351       45.0    1331.0    gray       30000.0  
3       yv1612tb4f1310987       41.0   14282.0   white       27750.0  
4       wba6b2c57ed129731       43.0    2641.0    gray       67000.0  
...                   ...        ...       ...     ...           ...  
558832  knalw4d4xf6019304       45.0   18255.0  silver       33000.0  
558833  3c6td5et6cg112407        5.0   54393.0   white       30800.0  
558834  5uxzw0c58cl668465       48.0   50561.0   black       34000.0  
558835  1n4al3ap0fc216050       38.0   16658.0   white       11100.0  
558836  1ftfw1et2eke87277       34.0   15008.0    gray       26700.0  

[510005 rows x 10 columns]

In [5]:
# choosing most value columns from the data base
columns_choosen = cars_formated.columns
columns_choosen = columns_choosen.drop('vin')

In [6]:
cars_formated.dtypes

year              int64
make             object
model            object
body             object
transmission     object
vin              object
condition       float64
odometer        float64
color            object
sellingprice    float64
dtype: object

In [7]:
# encoding from object to bool
one_hot_encoded_db = pd.get_dummies(cars_formated[columns_choosen])

one_hot_encoded_db

year  condition  odometer  sellingprice  make_acura  \
0       2015        5.0   16639.0       21500.0       False   
1       2015        5.0    9393.0       21500.0       False   
2       2014       45.0    1331.0       30000.0       False   
3       2015       41.0   14282.0       27750.0       False   
4       2014       43.0    2641.0       67000.0       False   
...      ...        ...       ...           ...         ...   
558832  2015       45.0   18255.0       33000.0       False   
558833  2012        5.0   54393.0       30800.0       False   
558834  2012       48.0   50561.0       34000.0       False   
558835  2015       38.0   16658.0       11100.0       False   
558836  2014       34.0   15008.0       26700.0       False   

        make_aston martin  make_audi  make_bentley  make_bmw  make_buick  ...  \
0                   False      False         False     False       False  ...   
1                   False      False         False     False       False  ...   
2                   False      False         False      True       False  ...   
3                   False      False         False     False       False  ...   
4                   False      False         False      True       False  ...   
...                   ...        ...           ...       ...         ...  ...   
558832              False      False         False     False       False  ...   
558833              False      False         False     False       False  ...   
558834              False      False         False      True       False  ...   
558835              False      False         False     False       False  ...   
558836              False      False         False     False       False  ...   

        color_lime  color_off-white  color_orange  color_pink  color_purple  \
0            False            False         False       False         False   
1            False            False         False       False         False   
2            False            False         False       False         False   
3            False            False         False       False         False   
4            False            False         False       False         False   
...            ...              ...           ...         ...           ...   
558832       False            False         False       False         False   
558833       False            False         False       False         False   
558834       False            False         False       False         False   
558835       False            False         False       False         False   
558836       False            False         False       False         False   

        color_red  color_silver  color_turquoise  color_white  color_yellow  
0           False         False            False         True         False  
1           False         False            False         True         False  
2           False         False            False        False         False  
3           False         False            False         True         False  
4           False         False            False        False         False  
...           ...           ...              ...          ...           ...  
558832      False          True            False        False         False  
558833      False         False            False         True         False  
558834      False         False            False        False         False  
558835      False         False            False         True         False  
558836      False         False            False        False         False  

[510005 rows x 893 columns]

# Spliting data to features and labels

In [8]:
features = one_hot_encoded_db.columns
# spliting to feature data base
# X = cars_formated[features]
features = features.drop('sellingprice')
features

Index(['year', 'condition', 'odometer', 'make_acura', 'make_aston martin',
       'make_audi', 'make_bentley', 'make_bmw', 'make_buick', 'make_cadillac',
       ...
       'color_lime', 'color_off-white', 'color_orange', 'color_pink',
       'color_purple', 'color_red', 'color_silver', 'color_turquoise',
       'color_white', 'color_yellow'],
      dtype='object', length=892)

In [9]:
# spliting to label
X = one_hot_encoded_db[features]
y = one_hot_encoded_db['sellingprice']
X

year  condition  odometer  make_acura  make_aston martin  make_audi  \
0       2015        5.0   16639.0       False              False      False   
1       2015        5.0    9393.0       False              False      False   
2       2014       45.0    1331.0       False              False      False   
3       2015       41.0   14282.0       False              False      False   
4       2014       43.0    2641.0       False              False      False   
...      ...        ...       ...         ...                ...        ...   
558832  2015       45.0   18255.0       False              False      False   
558833  2012        5.0   54393.0       False              False      False   
558834  2012       48.0   50561.0       False              False      False   
558835  2015       38.0   16658.0       False              False      False   
558836  2014       34.0   15008.0       False              False      False   

        make_bentley  make_bmw  make_buick  make_cadillac  ...  color_lime  \
0              False     False       False          False  ...       False   
1              False     False       False          False  ...       False   
2              False      True       False          False  ...       False   
3              False     False       False          False  ...       False   
4              False      True       False          False  ...       False   
...              ...       ...         ...            ...  ...         ...   
558832         False     False       False          False  ...       False   
558833         False     False       False          False  ...       False   
558834         False      True       False          False  ...       False   
558835         False     False       False          False  ...       False   
558836         False     False       False          False  ...       False   

        color_off-white  color_orange  color_pink  color_purple  color_red  \
0                 False         False       False         False      False   
1                 False         False       False         False      False   
2                 False         False       False         False      False   
3                 False         False       False         False      False   
4                 False         False       False         False      False   
...                 ...           ...         ...           ...        ...   
558832            False         False       False         False      False   
558833            False         False       False         False      False   
558834            False         False       False         False      False   
558835            False         False       False         False      False   
558836            False         False       False         False      False   

        color_silver  color_turquoise  color_white  color_yellow  
0              False            False         True         False  
1              False            False         True         False  
2              False            False        False         False  
3              False            False         True         False  
4              False            False        False         False  
...              ...              ...          ...           ...  
558832          True            False        False         False  
558833         False            False         True         False  
558834         False            False        False         False  
558835         False            False         True         False  
558836         False            False        False         False  

[510005 rows x 892 columns]

# Building first model

In [10]:
# importing models
from sklearn.tree import DecisionTreeRegressor
# import mae
from sklearn.metrics import mean_absolute_error
# import db split to validate data
from sklearn.model_selection import train_test_split

In [11]:
# define model. Specify a number for random_state to ensure same results each run
car_model = DecisionTreeRegressor(random_state=1)

# fit model
car_model.fit(X, y)

DecisionTreeRegressor(random_state=1)

In [12]:
# predicting price
pred_car_price = car_model.predict(X)
# checking mae
mae_pred = mean_absolute_error(y,pred_car_price)
mae_pred

0.30909500887246205

In [13]:
# % of the mean
mae_pred_proc = (mae_pred/y.mean())*100
mae_pred_proc

0.0022624299536692224

**Observation: MAE is very low, only `0.309`, it represents just `0.002%` of the average car price.**

# Validating model

In [14]:
# building new model

# spliting data to train and validation
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)
# define new car model
car_model_with_validation = DecisionTreeRegressor(random_state = 1)
# fiting model 
car_model_with_validation.fit(train_X, train_y)
# predicting price
pred_car_price_with_validation = car_model_with_validation.predict(val_X)

In [15]:
# checking mae
mae_pred_with_validation = mean_absolute_error(val_y, pred_car_price_with_validation)
mae_pred_with_validation

2003.383821430252

In [16]:
# % of the mean
mae_pred_with_validation_proc = (mae_pred_with_validation/y.mean())*100
mae_pred_with_validation_proc

14.663826448813053

**Observation: MAE is much bigger than without validation `2003.38`, it represents just `14.66%` of the average car price.**

# Underfitting and Overfitting

In [17]:
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state = 1)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

In [18]:
# compare MAE with differing values of max_leaf_nodes
for max_leaf_nodes in [5, 50, 500, 5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print("Max leaf nodes: %d  \t\t Mean Absolute Error:  %d" %(max_leaf_nodes, my_mae))

Max leaf nodes: 5  		 Mean Absolute Error:  4838
Max leaf nodes: 50  		 Mean Absolute Error:  3726
Max leaf nodes: 500  		 Mean Absolute Error:  2519
Max leaf nodes: 5000  		 Mean Absolute Error:  1750


In [19]:
my_mae_proc = (my_mae  /y.mean())*100
my_mae_proc

12.816139158053891

**Observation: From the `max_leaf_nodes = [5, 50, 500, 5000]`, the best MAE occurs when `max_leaf_nodes` is equal to `5000`, with a value of `1750`, representing just `12.82%`.**

# Creating random forest

In [20]:
# importing random forest
from sklearn.ensemble import RandomForestRegressor

# define random model
forest_model = RandomForestRegressor(random_state=1)
# fiting model
forest_model.fit(train_X, train_y)
# predicting 
car_price_forest_pred = forest_model.predict(val_X)
# compare car_price_forest_pred with val_y in mae
mae_random_forest = mean_absolute_error(val_y, car_price_forest_pred)

KeyboardInterrupt: 

In [ ]:
mae_random_forest_proc = (mae_random_forest/y.mean())*100
mae_random_forest_proc

**Observation: RandomForestRegressor have the best from previous method MAE equal `1515`,representing just `11.1%`.**

# Create Linear Regression

In [22]:
# importing Linear Regression
from sklearn.linear_model import LinearRegression

# creating Linear Regression model
linear_regression_model = LinearRegression()

# fiting model
linear_regression_model.fit(train_X, train_y)

# predicting price
pred_price_linear_regression = linear_regression_model.predict(val_X)

# compare car_price_forest_pred with val_y in mae
mae_linear_regression = mean_absolute_error(val_y, pred_price_linear_regression)
mae_linear_regression

2615.091969767928

In [23]:
mae_linear_regression_proc = (mae_linear_regression/y.mean())*100
mae_linear_regression_proc

19.141242123531157

In [ ]:
comparision_table = pd.DataFrame([
    ['MAE', mae_pred, mae_pred_with_validation, my_mae,mae_random_forest],
    ['MAE of mean %', mae_pred, mae_pred_with_validation_proc, my_mae_proc, mae_random_forest_proc]
], columns = ['','Full fill','Validadion','5000 leaf','Random Forest'])
comparision_table